src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Serialization/ConvertersForUnsupportedTypesFunctionalTests.cs

```diff
@@ -0,0 +1,100 @@
+
+            T deserializedValue = JsonSerializer.Deserialize<T>(json, options);
+            Assert.Equal(value, deserializedValue);
+
+            json = JsonSerializer.Serialize(wrappedValue, options);
+            Assert.Equal($"{{\"Property\":{expectedValueAsString}}}", json);
+            ClassWithProperty<T> deserializedWrappedValue = JsonSerializer.Deserialize<ClassWithProperty<T>>(json, options);
+            Assert.Equal(wrappedValue.Property, deserializedWrappedValue.Property);
+        }
+
+        public static IEnumerable<object[]> GetTestData()
+        {
+            yield return GetTestCase("0", Int128.Zero);
+            yield return GetTestCase("-1", new Int128(ulong.MaxValue, ulong.MaxValue));
+            yield return GetTestCase(Int128.MaxValue.ToString(), Int128.MaxValue);
+            yield return GetTestCase(ulong.MaxValue.ToString(), new Int128(0UL, ulong.MaxValue));
+            yield return GetTestCase(ulong.MaxValue.ToString() + "173", new Int128(0UL, ulong.MaxValue) * 1000 + 173);
+
+            yield return GetTestCase("0", UInt128.Zero);
+            yield return GetTestCase(UInt128.MaxValue.ToString(), UInt128.MaxValue);
+            yield return GetTestCase(ulong.MaxValue.ToString(), new UInt128(0UL, ulong.MaxValue));
+            yield return GetTestCase(ulong.MaxValue.ToString() + "173", new UInt128(0UL, ulong.MaxValue) * 1000 + 173);
+
+            yield return GetTestCase("0", Half.Zero);
+            yield return GetTestCase(Half.MaxValue.ToString(), Half.MaxValue);
+            yield return GetTestCase(Half.MinValue.ToString(), Half.MinValue);
+            yield return GetTestCase(((Half)1.45f).ToString(), (Half)1.45f);
+
+            static object[] GetTestCase(string expectedValue, object value) => new object[] { expectedValue, value };
+        }
+
+        internal class ClassWithProperty<T>
+        {
+            public T Property { get; set; }
+        }
+
+        internal class HalfConverter : SimpleConverter<Half>
+        {
+            public override Half Parse(string value) => Half.Parse(value);
+        }
+
+        internal class UInt128Converter : SimpleConverter<UInt128>
+        {
+            public override UInt128 Parse(string value) => UInt128.Parse(value);
+        }
+
+        internal class Int128Converter : SimpleConverter<Int128>
+        {
+            public override Int128 Parse(string value) => Int128.Parse(value);
+        }
+
+        internal abstract class SimpleConverter<T> : JsonConverter<T>
+        {
+            public abstract T Parse(string value);
+
+            public override T Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+            {
+                if (reader.HasValueSequence)
+                {
+                    return Parse(Encoding.UTF8.GetString(reader.ValueSequence));
+                }
+                else
+                {
+                    return Parse(Encoding.UTF8.GetString(reader.ValueSpan));
```

Comment:
In the interest of validating feasibility, can this converter be written in a way that avoids the intermediate string allocations?

Reply:
I'll need to convert to UTF-16 either way since none of these have UTF-8 overloads. I could possibly create char buffer on the stack with stackalloc but then ReadOnlySequence gets relatively complicated and I have to put some arbitrary numbers for guessed max size of the buffer. I think for simplicity better to keep it as is...